File: src/whp/cpuid.cpp
#include "cpuid.h"

#include "hv_enlightenment.h"

#include <Windows.h>
#include <intrin.h>

#include <atomic>
#include <vector>

namespace tinyvmm::whp {

namespace {

constexpr std::uint32_t kBit(unsigned i) noexcept { return 1u << i; }

// Feature bits we care about.
constexpr std::uint32_t kEcx_TscDeadline = kBit(24);   // CPUID.01H:ECX[24]
constexpr std::uint32_t kEcx_Hypervisor  = kBit(31);   // CPUID.01H:ECX[31]
constexpr std::uint32_t kEax_Arat        = kBit(2);    // CPUID.06H:EAX[2]
constexpr std::uint32_t kEdx_InvariantTsc = kBit(8);   // CPUID.80000007H:EDX[8]

// Runtime knob: when true, the dynamic CPUID handler masks off
// CPUID.01H:ECX[24] so the guest never sees TSC-deadline. Used as a fallback
// when WHP rejects WRMSR to 0x6E0 even with TSC-deadline in the static
// CPUID result list.
std::atomic<bool> g_hide_tsc_deadline{false};

// Hyper-V max-hypervisor-leaf. Must be >= 0x40000005 to clear
// `ms_hyperv_platform()`'s HYPERV_CPUID_MIN/MAX bracket. We expose six
// leaves (0x40000000..0x40000005) with the high leaves zeroed.
constexpr std::uint32_t kHvMaxLeaf = 0x40000006u;

// Compute the host TSC frequency, in Hz. Tries CPUID.15h on the host first,
// then CPUID.16h, then falls back to a short QueryPerformanceCounter-based
// calibration. Result is cached.
std::uint64_t MeasureTscHz() noexcept {
    int regs[4];
    __cpuid(regs, 0x15);
    const std::uint32_t den = static_cast<std::uint32_t>(regs[0]);
    const std::uint32_t num = static_cast<std::uint32_t>(regs[1]);
    const std::uint32_t ccc = static_cast<std::uint32_t>(regs[2]);
    if (den != 0 && num != 0 && ccc != 0) {
        return (static_cast<std::uint64_t>(ccc) * num) / den;
    }
    __cpuid(regs, 0x16);
    const std::uint32_t base_mhz = static_cast<std::uint32_t>(regs[0]);
    if (base_mhz != 0) {
        return static_cast<std::uint64_t>(base_mhz) * 1'000'000ull;
    }
    // Last resort: 50 ms calibration against QPC.
    LARGE_INTEGER qpc_freq{};
    LARGE_INTEGER qpc_a{};
    LARGE_INTEGER qpc_b{};
    QueryPerformanceFrequency(&qpc_freq);
    QueryPerformanceCounter(&qpc_a);
    const std::uint64_t tsc_a = __rdtsc();
    Sleep(50);
    QueryPerformanceCounter(&qpc_b);
    const std::uint64_t tsc_b = __rdtsc();
    const double secs =
        static_cast<double>(qpc_b.QuadPart - qpc_a.QuadPart) /
        static_cast<double>(qpc_freq.QuadPart);
    if (secs <= 0.0) return 2'000'000'000ull;  // conservative fallback
    return static_cast<std::uint64_t>(
        static_cast<double>(tsc_b - tsc_a) / secs);
}

std::uint64_t GetTscHz() noexcept {
    static const std::uint64_t cached = MeasureTscHz();
    return cached;
}

}  // namespace

CpuidResult ResolveCpuid(std::uint32_t leaf,
                         std::uint32_t subleaf,
                         std::uint32_t default_eax,
                         std::uint32_t default_ebx,
                         std::uint32_t default_ecx,
                         std::uint32_t default_edx,
                         const CpuidContext& vctx) noexcept {
    CpuidResult r{default_eax, default_ebx, default_ecx, default_edx};

    // Clamp / normalize the per-vCPU context defensively. Callers should
    // always pass values in range, but ResolveCpuid is also used directly
    // from --cpuid-test which exercises edge cases.
    const std::uint32_t vcpu_index =
        (vctx.vcpu_index < 256) ? vctx.vcpu_index : 0;
    const std::uint32_t vcpu_count =
        (vctx.vcpu_count >= 1 && vctx.vcpu_count <= 256)
            ? vctx.vcpu_count : 1;
    // Number of bits to shift x2APIC ID right to reach the package level
    // from the Core level. We always advertise the kMaxVcpus ceiling (5
    // bits covers 0..31) so the topology shape is stable regardless of
    // how many vCPUs were actually spawned. This is the same trick KVM
    // and most other VMMs use; the actual count of online CPUs comes from
    // the MADT, not from CPUID.
    constexpr std::uint32_t kCoreShift = 5;  // ceil(log2(32))

    switch (leaf) {
        case 0x00000000u: {
            // Ensure max-standard-leaf is high enough for leaves 0x15 / 0x16
            // (which Linux uses to skip PIT calibration) AND 0x1F / 0x0B
            // (x2APIC extended topology, which our handler synthesizes
            // below). Modern hosts already report >= 0x1F; this is just a
            // safety raise.
            if (r.eax < 0x1Fu) r.eax = 0x1Fu;
            break;
        }

        case 0x00000001u: {
            // Standard feature flags. tsc-deadline lets Linux program the LAPIC
            // timer as a TSC deadline (no LAPIC frequency calibration). Set the
            // hypervisor-present bit so Linux reads the 0x40000000h vendor
            // string -- only a log line results, but it's informative and stops
            // the kernel from treating the host like bare metal.
            if (!g_hide_tsc_deadline.load(std::memory_order_relaxed)) {
                r.ecx |= kEcx_TscDeadline;
            } else {
                r.ecx &= ~kEcx_TscDeadline;
            }
            r.ecx |= kEcx_Hypervisor;

            // EBX layout for leaf 0x01:
            //   [7:0]   brand index            -- preserve host
            //   [15:8]  CLFLUSH line size / 8  -- preserve host
            //   [23:16] max logical procs / package -- we own this
            //   [31:24] initial APIC ID (8-bit) -- we own this
            //
            // Without rewriting [31:24], MADT says vCPU N has APIC ID N but
            // CPUID returns the *host* CPU's APIC ID (whatever core happens
            // to be running this vCPU thread), and Linux logs
            // `[Firmware Bug]: CPU N: APIC ID mismatch`. Linux also uses
            // [23:16] for "smp_num_siblings" / max-logical-per-package as
            // a legacy fallback before reading 0x0B/0x1F, so we set that
            // to vcpu_count for coherence.
            r.ebx = (r.ebx & 0x0000FFFFu) |
                    ((vcpu_count & 0xFFu) << 16) |
                    ((vcpu_index & 0xFFu) << 24);
            break;
        }

        case 0x00000006u: {
            // Thermal / power leaf. Bit 2 of EAX is ARAT: APIC timer is always
            // running (doesn't stop in deep C-states). Setting this avoids
            // Linux falling back to slower clock sources for hrtimers.
            r.eax |= kEax_Arat;
            break;
        }

        case 0x0000000Bu:
        case 0x0000001Fu: {
            // x2APIC extended topology enumeration. Leaf 0x1F is the v2
            // version of 0x0B; both share the same shape for the levels
            // we model (SMT and Core). Linux's
            // `arch/x86/kernel/cpu/topology_common.c::topo_get_cpuid()`
            // walks subleaves until ECX[15:8] returns 0 (invalid). For
            // every subleaf the per-vCPU x2APIC ID is reported in EDX,
            // which is what fixes the APIC-ID-mismatch firmware-bug
            // message.
            //
            // Register layout per subleaf:
            //   EAX [4:0]   bits to shift x2APIC ID right to reach the
            //               next-level ID (typed by ECX[15:8])
            //   EBX [15:0]  number of logical processors at this level
            //   ECX [7:0]   subleaf number (echoed back)
            //   ECX [15:8]  level type (0=Invalid, 1=SMT, 2=Core,
            //               3=Module, 4=Tile, 5=Die)
            //   EDX         32-bit x2APIC ID of the current logical proc
            constexpr std::uint8_t kLevelInvalid = 0;
            constexpr std::uint8_t kLevelSmt     = 1;
            constexpr std::uint8_t kLevelCore    = 2;

            std::uint32_t eax = 0, ebx = 0, level = kLevelInvalid;
            switch (subleaf) {
                case 0: {
                    // SMT level: 1 thread per core (no hyperthreading).
                    eax   = 0;
                    ebx   = 1;
                    level = kLevelSmt;
                    break;
                }
                case 1: {
                    // Core level: vcpu_count cores per package. We always
                    // shift by kCoreShift bits so the topology shape is
                    // stable across --vcpus values; package ID is
                    // x2apic_id >> kCoreShift.
                    eax   = kCoreShift;
                    ebx   = vcpu_count;
                    level = kLevelCore;
                    break;
                }
                default: {
                    // Invalid-level terminator. EAX/EBX zero, ECX echoes
                    // back the subleaf with level=Invalid.
                    eax   = 0;
                    ebx   = 0;
                    level = kLevelInvalid;
                    break;
                }
            }
            r.eax = eax;
            r.ebx = ebx & 0xFFFFu;
            r.ecx = (subleaf & 0xFFu) | (static_cast<std::uint32_t>(level) << 8);
            r.edx = vcpu_index;
            break;
        }

        case 0x00000015u: {
            // TSC / core-crystal frequency ratio. Encode as
            //   TSC_Hz = ECX * EBX / EAX
            // with EBX=EAX=1 so ECX is the literal TSC frequency. Linux's
            // arch/x86/kernel/tsc.c uses this to skip PIT calibration entirely
            // when ECX != 0 and (EBX,EAX) != 0.
            const std::uint64_t tsc_hz = GetTscHz();
            r.eax = 1;
            r.ebx = 1;
            r.ecx = (tsc_hz <= 0xFFFFFFFFull)
                ? static_cast<std::uint32_t>(tsc_hz)
                : 0xFFFFFFFFu;
            r.edx = 0;
            break;
        }

        case 0x00000016u: {
            // CPU base / max / bus frequency hint in MHz.
            const std::uint64_t tsc_hz = GetTscHz();
            const std::uint32_t base_mhz =
                static_cast<std::uint32_t>(tsc_hz / 1'000'000ull);
            r.eax = base_mhz;       // base
            r.ebx = base_mhz;       // max (we don't model turbo)
            r.ecx = 100;            // bus, conventional 100 MHz reference
            r.edx = 0;
            break;
        }

        case 0x80000007u: {
            // Power management leaf. Bit 8 of EDX is invariant_tsc. We set it
            // because the underlying host has invariant TSC on every CPU we
            // care about, and Linux uses this bit to pick the TSC clocksource.
            r.edx |= kEdx_InvariantTsc;
            break;
        }

        case 0x40000000u: {
            // Max hypervisor leaf + 12-byte vendor "Microsoft Hv" packed
            // across EBX:ECX:EDX. Linux's `ms_hyperv_platform()` requires
            // BOTH this exact vendor string AND a max leaf in
            // [HYPERV_CPUID_MIN, HYPERV_CPUID_MAX] = [0x40000005, 0x4000ffff]
            // before it will treat us as Hyper-V and route through the
            // Reference TSC page setup.
            r.eax = kHvMaxLeaf;
            std::uint32_t ebx = 0, ecx = 0, edx = 0;
            GetHvVendorEbxEcxEdx(&ebx, &ecx, &edx);
            r.ebx = ebx;
            r.ecx = ecx;
            r.edx = edx;
            break;
        }

        case 0x40000001u: {
            // Hyper-V interface signature "Hv#1". `ms_hyperv_platform()`
            // checks this with memcmp("Hv#1", &eax, 4); any deviation
            // disables the Hyper-V path entirely.
            r.eax = GetHvInterfaceEax();
            r.ebx = 0;
            r.ecx = 0;
            r.edx = 0;
            break;
        }

        case 0x40000002u: {
            // Hypervisor system identity (build/major/minor/service pack).
            // Linux logs the version but does not gate behavior on it.
            r.eax = 0; r.ebx = 0; r.ecx = 0; r.edx = 0;
            break;
        }

        case 0x40000003u: {
            // Feature flags. EAX advertises the four enlightenments we
            // implement (hypercall page, VP_INDEX, Reference TSC,
            // TSC-invariant control). EBX/ECX/EDX are zero so Linux skips
            // hypercall-only features (FREQUENCY_MSRS, SynIC, synthetic
            // timers, partition-reference-counter, etc.).
            r.eax = kHvFeaturesAdvertised;
            r.ebx = 0;
            r.ecx = 0;
            r.edx = 0;
            break;
        }

        case 0x40000004u:
        case 0x40000005u:
        case 0x40000006u: {
            // Recommended enlightenment hints (0x40000004) / implementation
            // limits (0x40000005) / hardware features (0x40000006). All
            // zero -- no PV-TLB-flush, no PV-IPI, no SynIC, no recommended
            // hypercall paths. Linux honors this by skipping the entire
            // Hyper-V fast-IPI / cluster-IPI plumbing.
            r.eax = 0; r.ebx = 0; r.ecx = 0; r.edx = 0;
            break;
        }

        default: {
            if (leaf >= 0x40000007u && leaf <= 0x400000FFu) {
                // Reserved hypervisor leaves -- zero everything so the guest
                // doesn't pick up stale host CPUID fragments here.
                r.eax = 0;
                r.ebx = 0;
                r.ecx = 0;
                r.edx = 0;
            }
            break;
        }
    }

    return r;
}

void SetHideTscDeadline(bool hide) noexcept {
    g_hide_tsc_deadline.store(hide, std::memory_order_relaxed);
}
bool GetHideTscDeadline() noexcept {
    return g_hide_tsc_deadline.load(std::memory_order_relaxed);
}

std::uint64_t GetCachedTscHz() noexcept {
    return GetTscHz();
}

std::vector<WHV_X64_CPUID_RESULT> BuildStaticCpuidResultList(
    bool hide_tsc_deadline) {
    auto host = [](std::uint32_t leaf,
                   std::uint32_t subleaf = 0) -> CpuidResult {
        int regs[4] = {};
        __cpuidex(regs, static_cast<int>(leaf), static_cast<int>(subleaf));
        return {static_cast<std::uint32_t>(regs[0]),
                static_cast<std::uint32_t>(regs[1]),
                static_cast<std::uint32_t>(regs[2]),
                static_cast<std::uint32_t>(regs[3])};
    };

    auto entry = [](std::uint32_t leaf, const CpuidResult& r,
                    std::uint32_t /*subleaf*/ = 0) {
        WHV_X64_CPUID_RESULT e = {};
        e.Function = leaf;
        e.Eax = r.eax;
        e.Ebx = r.ebx;
        e.Ecx = r.ecx;
        e.Edx = r.edx;
        return e;
    };

    std::vector<WHV_X64_CPUID_RESULT> list;
    list.reserve(16);

    // For each customized leaf, start with the host's CPUID value and apply
    // the same overrides as the runtime path. This keeps the static
    // architectural model WHP sees consistent with what the guest reads at
    // runtime via the CPUID exit handler.
    //
    // The CpuidResultList is per-partition (one entry per leaf shared by
    // all vCPUs), so we encode the BSP (vcpu_index=0) values here. The
    // runtime CPUID exit handler always wins at execution time and
    // returns the actual per-vCPU values; this static list is only used
    // by WHP for internal architectural decisions taken at SetupPartition
    // time. Leaves 0x0B / 0x1F (per-vCPU topology) are intentionally
    // omitted -- their subleaf-shaped, vCPU-specific output cannot be
    // expressed in a static partition-wide table, so we let the runtime
    // path handle them every time.
    {
        CpuidResult r0 = host(0x00000000u);
        if (r0.eax < 0x1Fu) r0.eax = 0x1Fu;
        list.push_back(entry(0x00000000u, r0));
    }
    {
        CpuidResult r1 = host(0x00000001u);
        if (!hide_tsc_deadline) {
            r1.ecx |= kEcx_TscDeadline;
        } else {
            r1.ecx &= ~kEcx_TscDeadline;
        }
        r1.ecx |= kEcx_Hypervisor;
        // Static-model EBX[31:24] = 0 (BSP APIC ID), EBX[23:16] = 1
        // (single logical processor per package fallback). The runtime
        // exit handler overrides both with the actual vcpu_index and
        // vcpu_count.
        r1.ebx = (r1.ebx & 0x0000FFFFu) | (1u << 16) | (0u << 24);
        list.push_back(entry(0x00000001u, r1));
    }
    {
        CpuidResult r6 = host(0x00000006u);
        r6.eax |= kEax_Arat;
        list.push_back(entry(0x00000006u, r6));
    }
    {
        // 0x15: TSC / core-crystal frequency ratio, same as runtime path.
        const std::uint64_t tsc_hz = GetTscHz();
        CpuidResult r{};
        r.eax = 1;
        r.ebx = 1;
        r.ecx = (tsc_hz <= 0xFFFFFFFFull)
                    ? static_cast<std::uint32_t>(tsc_hz)
                    : 0xFFFFFFFFu;
        r.edx = 0;
        list.push_back(entry(0x00000015u, r));
    }
    {
        // 0x16: CPU base / max / bus frequency in MHz.
        const std::uint64_t tsc_hz = GetTscHz();
        const std::uint32_t base_mhz =
            static_cast<std::uint32_t>(tsc_hz / 1'000'000ull);
        CpuidResult r{base_mhz, base_mhz, 100u, 0u};
        list.push_back(entry(0x00000016u, r));
    }
    {
        CpuidResult r = host(0x80000007u);
        r.edx |= kEdx_InvariantTsc;
        list.push_back(entry(0x80000007u, r));
    }
    {
        // 0x40000000: hypervisor max-leaf + vendor "Microsoft Hv".
        std::uint32_t ebx = 0, ecx = 0, edx = 0;
        GetHvVendorEbxEcxEdx(&ebx, &ecx, &edx);
        CpuidResult r{kHvMaxLeaf, ebx, ecx, edx};
        list.push_back(entry(0x40000000u, r));
    }
    {
        // 0x40000001: hypervisor interface signature "Hv#1".
        CpuidResult r{GetHvInterfaceEax(), 0, 0, 0};
        list.push_back(entry(0x40000001u, r));
    }
    {
        // 0x40000002: hypervisor build/version, all zero.
        list.push_back(entry(0x40000002u, CpuidResult{0, 0, 0, 0}));
    }
    {
        // 0x40000003: features. EAX = HYPERCALL | VP_INDEX | REFERENCE_TSC
        //             | TSC_INVARIANT. Other registers zero.
        CpuidResult r{kHvFeaturesAdvertised, 0, 0, 0};
        list.push_back(entry(0x40000003u, r));
    }
    // Mask 0x40000004..0x40000006 (hints / impl limits / hw features) so
    // the guest's `ms_hyperv.hints` etc. read zero, suppressing PV TLB
    // flush, PV IPI, SynIC, and similar features we don't implement.
    for (std::uint32_t leaf = 0x40000004u; leaf <= 0x40000006u; ++leaf) {
        list.push_back(entry(leaf, CpuidResult{0, 0, 0, 0}));
    }

    return list;
}

}  // namespace tinyvmm::whp

File: tinyvmm-rs/src/whp/cpuid.rs
//! CPUID exit policy. Layers tinyvmm overrides (invariant TSC, ARAT, TSC
//! frequency via leaf 0x15/0x16, per-vCPU x2APIC topology, and the Hyper-V
//! vendor/feature leaves) on top of WHP's host-passthrough defaults. Port of
//! src/whp/cpuid.cpp.

use core::arch::x86_64::{__cpuid, __cpuid_count, _rdtsc};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::OnceLock;
use windows_sys::Win32::System::Hypervisor::WHV_X64_CPUID_RESULT;

#[derive(Clone, Copy, Default)]
pub struct CpuidResult {
    pub eax: u32,
    pub ebx: u32,
    pub ecx: u32,
    pub edx: u32,
}

pub struct CpuidContext {
    pub vcpu_index: u32,
    pub vcpu_count: u32,
}

const ECX_TSC_DEADLINE: u32 = 1 << 24;
const ECX_HYPERVISOR: u32 = 1 << 31;
const EAX_ARAT: u32 = 1 << 2;
const EDX_INVARIANT_TSC: u32 = 1 << 8;

const HV_MAX_LEAF: u32 = 0x4000_0006;

// Hyper-V CPUID.40000003H:EAX feature bits we advertise.
const HV_FEATURE_HYPERCALL: u32 = 1 << 5;
const HV_FEATURE_VP_INDEX: u32 = 1 << 6;
const HV_FEATURE_REFERENCE_TSC: u32 = 1 << 9;
const HV_FEATURE_TSC_INVARIANT: u32 = 1 << 15;
pub const HV_FEATURES_ADVERTISED: u32 =
    HV_FEATURE_HYPERCALL | HV_FEATURE_VP_INDEX | HV_FEATURE_REFERENCE_TSC | HV_FEATURE_TSC_INVARIANT;

static HIDE_TSC_DEADLINE: AtomicBool = AtomicBool::new(false);

pub fn set_hide_tsc_deadline(hide: bool) {
    HIDE_TSC_DEADLINE.store(hide, Ordering::Relaxed);
}

/// "Microsoft Hv" packed across three little-endian dwords (EBX:ECX:EDX).
pub fn hv_vendor() -> (u32, u32, u32) {
    let pack = |b: [u8; 4]| {
        (b[0] as u32) | ((b[1] as u32) << 8) | ((b[2] as u32) << 16) | ((b[3] as u32) << 24)
    };
    (
        pack(*b"Micr"),
        pack(*b"osof"),
        pack(*b"t Hv"),
    )
}

/// Hyper-V interface signature "Hv#1" packed into CPUID.40000001H:EAX.
pub fn hv_interface_eax() -> u32 {
    let b = *b"Hv#1";
    (b[0] as u32) | ((b[1] as u32) << 8) | ((b[2] as u32) << 16) | ((b[3] as u32) << 24)
}

fn measure_tsc_hz() -> u64 {
    let r = __cpuid(0x15);
    let den = r.eax;
    let num = r.ebx;
    let ccc = r.ecx;
    if den != 0 && num != 0 && ccc != 0 {
        return (ccc as u64 * num as u64) / den as u64;
    }
    let r = __cpuid(0x16);
    let base_mhz = r.eax;
    if base_mhz != 0 {
        return base_mhz as u64 * 1_000_000;
    }
    // Last resort: ~50 ms calibration against the monotonic clock.
    let t0 = std::time::Instant::now();
    // SAFETY: _rdtsc just reads the timestamp counter; always available on x86_64.
    let tsc_a = unsafe { _rdtsc() };
    std::thread::sleep(std::time::Duration::from_millis(50));
    let tsc_b = unsafe { _rdtsc() };
    let secs = t0.elapsed().as_secs_f64();
    if secs <= 0.0 {
        return 2_000_000_000;
    }
    ((tsc_b - tsc_a) as f64 / secs) as u64
}

pub fn cached_tsc_hz() -> u64 {
    static CACHED: OnceLock<u64> = OnceLock::new();
    *CACHED.get_or_init(measure_tsc_hz)
}

pub fn resolve_cpuid(
    leaf: u32,
    subleaf: u32,
    default_eax: u32,
    default_ebx: u32,
    default_ecx: u32,
    default_edx: u32,
    ctx: &CpuidContext,
) -> CpuidResult {
    let mut r = CpuidResult {
        eax: default_eax,
        ebx: default_ebx,
        ecx: default_ecx,
        edx: default_edx,
    };

    let vcpu_index = if ctx.vcpu_index < 256 { ctx.vcpu_index } else { 0 };
    let vcpu_count = if ctx.vcpu_count >= 1 && ctx.vcpu_count <= 256 {
        ctx.vcpu_count
    } else {
        1
    };
    const CORE_SHIFT: u32 = 5; // ceil(log2(32))

    match leaf {
        0x0000_0000 => {
            if r.eax < 0x1F {
                r.eax = 0x1F;
            }
        }
        0x0000_0001 => {
            if !HIDE_TSC_DEADLINE.load(Ordering::Relaxed) {
                r.ecx |= ECX_TSC_DEADLINE;
            } else {
                r.ecx &= !ECX_TSC_DEADLINE;
            }
            r.ecx |= ECX_HYPERVISOR;
            r.ebx = (r.ebx & 0x0000_FFFF)
                | ((vcpu_count & 0xFF) << 16)
                | ((vcpu_index & 0xFF) << 24);
        }
        0x0000_0006 => {
            r.eax |= EAX_ARAT;
        }
        0x0000_000B | 0x0000_001F => {
            const LEVEL_INVALID: u32 = 0;
            const LEVEL_SMT: u32 = 1;
            const LEVEL_CORE: u32 = 2;
            let (eax, ebx, level) = match subleaf {
                0 => (0u32, 1u32, LEVEL_SMT),
                1 => (CORE_SHIFT, vcpu_count, LEVEL_CORE),
                _ => (0, 0, LEVEL_INVALID),
            };
            r.eax = eax;
            r.ebx = ebx & 0xFFFF;
            r.ecx = (subleaf & 0xFF) | (level << 8);
            r.edx = vcpu_index;
        }
        0x0000_0015 => {
            let tsc_hz = cached_tsc_hz();
            r.eax = 1;
            r.ebx = 1;
            r.ecx = if tsc_hz <= 0xFFFF_FFFF {
                tsc_hz as u32
            } else {
                0xFFFF_FFFF
            };
            r.edx = 0;
        }
        0x0000_0016 => {
            let tsc_hz = cached_tsc_hz();
            let base_mhz = (tsc_hz / 1_000_000) as u32;
            r.eax = base_mhz;
            r.ebx = base_mhz;
            r.ecx = 100;
            r.edx = 0;
        }
        0x8000_0007 => {
            r.edx |= EDX_INVARIANT_TSC;
        }
        0x4000_0000 => {
            r.eax = HV_MAX_LEAF;
            let (ebx, ecx, edx) = hv_vendor();
            r.ebx = ebx;
            r.ecx = ecx;
            r.edx = edx;
        }
        0x4000_0001 => {
            r.eax = hv_interface_eax();
            r.ebx = 0;
            r.ecx = 0;
            r.edx = 0;
        }
        0x4000_0002 => {
            r = CpuidResult::default();
        }
        0x4000_0003 => {
            r.eax = HV_FEATURES_ADVERTISED;
            r.ebx = 0;
            r.ecx = 0;
            r.edx = 0;
        }
        0x4000_0004..=0x4000_0006 => {
            r = CpuidResult::default();
        }
        other => {
            if (0x4000_0007..=0x4000_00FF).contains(&other) {
                r = CpuidResult::default();
            }
        }
    }

    r
}

fn host_cpuid(leaf: u32, subleaf: u32) -> CpuidResult {
    let r = __cpuid_count(leaf, subleaf);
    CpuidResult {
        eax: r.eax,
        ebx: r.ebx,
        ecx: r.ecx,
        edx: r.edx,
    }
}

fn entry(leaf: u32, r: &CpuidResult) -> WHV_X64_CPUID_RESULT {
    WHV_X64_CPUID_RESULT {
        Function: leaf,
        Reserved: [0; 3],
        Eax: r.eax,
        Ebx: r.ebx,
        Ecx: r.ecx,
        Edx: r.edx,
    }
}

/// Static CPUID result list mirroring `BuildStaticCpuidResultList`.
pub fn build_static_cpuid_result_list(hide_tsc_deadline: bool) -> Vec<WHV_X64_CPUID_RESULT> {
    let mut list = Vec::with_capacity(16);

    {
        let mut r0 = host_cpuid(0x0000_0000, 0);
        if r0.eax < 0x1F {
            r0.eax = 0x1F;
        }
        list.push(entry(0x0000_0000, &r0));
    }
    {
        let mut r1 = host_cpuid(0x0000_0001, 0);
        if !hide_tsc_deadline {
            r1.ecx |= ECX_TSC_DEADLINE;
        } else {
            r1.ecx &= !ECX_TSC_DEADLINE;
        }
        r1.ecx |= ECX_HYPERVISOR;
        r1.ebx = (r1.ebx & 0x0000_FFFF) | (1 << 16);
        list.push(entry(0x0000_0001, &r1));
    }
    {
        let mut r6 = host_cpuid(0x0000_0006, 0);
        r6.eax |= EAX_ARAT;
        list.push(entry(0x0000_0006, &r6));
    }
    {
        let tsc_hz = cached_tsc_hz();
        let r = CpuidResult {
            eax: 1,
            ebx: 1,
            ecx: if tsc_hz <= 0xFFFF_FFFF {
                tsc_hz as u32
            } else {
                0xFFFF_FFFF
            },
            edx: 0,
        };
        list.push(entry(0x0000_0015, &r));
    }
    {
        let tsc_hz = cached_tsc_hz();
        let base_mhz = (tsc_hz / 1_000_000) as u32;
        let r = CpuidResult {
            eax: base_mhz,
            ebx: base_mhz,
            ecx: 100,
            edx: 0,
        };
        list.push(entry(0x0000_0016, &r));
    }
    {
        let mut r = host_cpuid(0x8000_0007, 0);
        r.edx |= EDX_INVARIANT_TSC;
        list.push(entry(0x8000_0007, &r));
    }
    {
        let (ebx, ecx, edx) = hv_vendor();
        let r = CpuidResult {
            eax: HV_MAX_LEAF,
            ebx,
            ecx,
            edx,
        };
        list.push(entry(0x4000_0000, &r));
    }
    {
        let r = CpuidResult {
            eax: hv_interface_eax(),
            ebx: 0,
            ecx: 0,
            edx: 0,
        };
        list.push(entry(0x4000_0001, &r));
    }
    list.push(entry(0x4000_0002, &CpuidResult::default()));
    {
        let r = CpuidResult {
            eax: HV_FEATURES_ADVERTISED,
            ebx: 0,
            ecx: 0,
            edx: 0,
        };
        list.push(entry(0x4000_0003, &r));
    }
    for leaf in 0x4000_0004..=0x4000_0006u32 {
        list.push(entry(leaf, &CpuidResult::default()));
    }

    list
}
